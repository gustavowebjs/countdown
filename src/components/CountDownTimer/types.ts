export type TimeFormat = {
  minutes: number;
  seconds: number;
};

export type TimerProps = {
  timeFormat: TimeFormat;
  active?: boolean;
  speed?: number;
};
