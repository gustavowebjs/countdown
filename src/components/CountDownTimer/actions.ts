type CurrentlyTimerAction = {
  mins: number;
  secs: number;
};

type TimerAction = {
  minutes: number;
  seconds: number;
};

export const handleTimerText = (
  currently: CurrentlyTimerAction,
  time: TimerAction
): string => {
  const { mins, secs } = currently;
  const { minutes, seconds } = time;

  const currentlyTime = mins * 60 + secs;
  const timeToEnd = minutes * 60 + seconds;

  if (currentlyTime === 0) {
    return 'TimeÂ´s up!';
  }

  if (currentlyTime < timeToEnd / 2) {
    return `More than halfway there!`;
  }

  return '';
};

export const verifyTimeBreaks = (time: CurrentlyTimerAction): string => {
  const { mins, secs } = time;

  const currentlyTime = mins * 60 + secs;

  if (currentlyTime == 0) {
    return 'timesup';
  }

  if (currentlyTime <= 10) {
    return '10seconds';
  }

  if (currentlyTime <= 20) {
    return '20seconds';
  }

  return '';
};

export const formatTimeToRender = (mins: number, secs: number): string => {
  return `${mins.toString().padStart(2, '0')}:${secs
    .toString()
    .padStart(2, '0')}`;
};
