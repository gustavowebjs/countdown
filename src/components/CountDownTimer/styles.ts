import styled, { css } from 'styled-components';

interface TimerStyledProps {
  active: boolean;
}

export const Container = styled.div<TimerStyledProps>`
  display: flex;
  flex-direction: column;
  align-items: center;

  > p {
    font-size: 6.2rem;
    font-weight: bold;
    color: #ccc;

    ${({ active }) =>
      active === true &&
      css`
        color: ${({ theme }) => theme.colors.neutral.dark};
      `}
  }
`;
