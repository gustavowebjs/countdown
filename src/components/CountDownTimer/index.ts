import CountDownTimer from './component';

export default CountDownTimer;
