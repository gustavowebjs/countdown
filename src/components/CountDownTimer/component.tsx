import { useEffect, useState } from 'react';
import * as T from './types';
import * as S from './styles';
import {
  formatTimeToRender,
  handleTimerText,
  verifyTimeBreaks,
} from './actions';

const CountDownTimer = ({
  timeFormat,
  active = false,
  speed = 1000,
  onComplete,
}: T.TimerProps) => {
  const { minutes = 20, seconds = 60 } = timeFormat;
  const [timesUp, setTimesUp] = useState<boolean>(false);
  const [timerText, setTimerText] = useState<string>('');
  const [styleText, setStyleText] = useState<string>('');
  const [[mins, secs], setTime] = useState<[number, number]>([
    minutes,
    seconds,
  ]);

  const tick = (): void => {
    if (mins === 0 && secs === 0) {
      setTimesUp(true);
    } else if (secs === 0) {
      setTime([mins - 1, 59]);
    } else {
      setTime([mins, secs - 1]);
    }
  };

  useEffect(() => {
    if (active === true && !timesUp) {
      const text = handleTimerText({ mins, secs }, { minutes, seconds });
      setTimerText(text);

      const styleTextBreak = verifyTimeBreaks({ mins, secs });
      setStyleText(styleTextBreak);

      const timerId = setInterval(() => tick(), speed);
      return () => clearInterval(timerId);
    }
  });

  useEffect(() => {
    setTime([minutes, seconds]);
    setTimesUp(false);
  }, [timeFormat]);

  useEffect(() => {
    if (timesUp) onComplete && onComplete();
  }, [timesUp]);

  return (
    <S.Container styleText={styleText} active={active}>
      <S.TimerText> {timerText}</S.TimerText>
      <p>{formatTimeToRender(mins, secs)}</p>
    </S.Container>
  );
};

export default CountDownTimer;
