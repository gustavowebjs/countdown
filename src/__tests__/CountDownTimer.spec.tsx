import { shallow } from 'enzyme';
import CountDownTimer from '../components/CountDownTimer';

describe('CountDownTimer Component', () => {
  it('Should Countdown component render correctly', () => {
    const wrapper = shallow(
      <CountDownTimer speed={1000} timeFormat={{ minutes: 20, seconds: 0 }} />
    );
    expect(wrapper.text()).toEqual(' 20:00');
    expect(wrapper).toMatchSnapshot();
  });
});
