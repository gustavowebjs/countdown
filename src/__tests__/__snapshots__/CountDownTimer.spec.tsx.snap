// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountDownTimer Component Should Countdown component render correctly 1`] = `
<styled.div
  active={false}
  styleText=""
>
  <styled.h3>
     
  </styled.h3>
  <p>
    20:00
  </p>
</styled.div>
`;
